ledger/streaming/publisher/src/sink.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::BlockPublisher;
+use futures::{Future, Stream, StreamExt};
+use mc_common::logger::Logger;
+use mc_crypto_keys::Ed25519Pair;
+use mc_ledger_streaming_api::{
+    make_subscribe_response, streaming_blocks::SubscribeResponse, BlockStreamComponents,
+};
+use mc_util_grpc::ConnectionUriGrpcioServer;
+use mc_util_uri::ConnectionUri;
+use std::sync::{Arc, RwLock};
+
+// A sink that consumes a block stream and publishes the results over gRPC.
+pub struct GrpcServerSink {
+    publisher: Arc<RwLock<BlockPublisher>>,
+    signer: Arc<RwLock<Ed25519Pair>>,
+    logger: Logger,
+}
+
+impl GrpcServerSink {
+    pub fn new(signer: Ed25519Pair, logger: Logger) -> Self {
+        Self {
+            publisher: Arc::new(RwLock::new(BlockPublisher::new(logger.clone()))),
+            signer: Arc::new(RwLock::new(signer)),
+            logger,
+        }
+    }
+
+    pub fn consume<'a>(
+        &mut self,
+        stream: impl Stream<Item = BlockStreamComponents> + 'a,
+    ) -> impl Future<Output = ()> + 'a {
+        let signer = self.signer.clone();
+        self.consume_protos(stream.map(move |data| {
+            make_subscribe_response(&data, &signer.read().unwrap())
+                .expect("Failed to create SubscribeResponse")
+        }))
+    }
```

Comment:
Should we wrap this one more time to return a stream?

Reply:
I'm not convinced a sink should implement Stream. It seems easier to fork/duplicate the stream ahead of this sink, than to make it propagate the stream.